libraries/abstractions/secure_sockets/test/iot_test_udp.c

```diff
@@ -498,6 +506,91 @@ static BaseType_t prvCheckRxTxBuffers( uint8_t * pucTxBuffer,
 }
 /*-----------------------------------------------------------*/
 
+/*
+ * @brief Test SendTo and RecvFrom echo server. Return pdFAIL if one of the operation fails, allowing retry.
+ * We expect the socket to not fail during retry.
+ */
+static BaseType_t prvSendRecvEchoTestHelper( Socket_t xSocket,
+                                             size_t xLength,
+                                             SocketsSockaddr_t * pxEchoServerAddress,
+                                             uint32_t ulTxCount )
+{
+    BaseType_t xResult;
+    BaseType_t xIntermResult;
+    uint8_t * pucTxBuffer = ( uint8_t * ) pcTxBuffer;
+    uint8_t * pucRxBuffer = ( uint8_t * ) pcRxBuffer;
+    SocketsSockaddr_t xReceiveAddress = { 0 };
+    uint32_t ulRetry;
+
+    xIntermResult = pdPASS;
+    xResult = pdPASS;
+
+    memset( pucTxBuffer, udptestTX_BUFFER_FILLER, udptestBUFFER_SIZE );
+
+    prvCreateTxData( ( char * ) pucTxBuffer,
+                     xLength,
+                     ulTxCount );
+
+    /* Since UDP does not guarantee data delivery, we will retry until there is a
+     * successful loop */
+    for( ulRetry = 0; ulRetry < udptestMAX_RETRY; ulRetry++ )
+    {
+        xIntermResult = prvSendHelper( xSocket,
+                                       pucTxBuffer,
+                                       xLength,
+                                       pxEchoServerAddress );
+
+        if( xIntermResult != pdPASS )
+        {
+            udptestPRINTF( ( "Tx %d data failed to send on try %d\r\n", ulTxCount, ulRetry ) );
+            continue;
```

Comment:
I notice you have vTaskDelay() for other cases below, but not this one.  Consider having one vTaskDelay() at the top of the loop when ulRetry > 0.

Reply:
Adding a delay on every loop may delay the test. Since we are not sure if adding delay (and how much) helps with higher chance of successful transmission, I'm inclined to add it only when loop failed.
Without delay at each loop, the platform with the highest fail rate is able to achieve 99.90% successful transmission.